### PR TITLE
Shrink Task.Delay when used without Cancellation

### PR DESCRIPTION
**Pull request by benaadams**
`Task.Delay` is commonly used without cancellation:

`Task Delay(int)` - API Port 22.6%
`Task Delay(TimeSpan)` - API Port 9.8%

`Task Delay(int, CancellationToken)` - API Port 5.6%
`Task Delay(TimeSpan, CancellationToken)` - API Port 9.3%

This shaves 24 bytes of allocation off that common path.

Using @SergeyTeplyakov's [ObjectLayoutInspector](https://github.com/SergeyTeplyakov/ObjectLayoutInspector) and local copies of the internal objects to make it easier to inspect.
```csharp
ObjectLayoutInspector.TypeLayout.PrintLayout<DelayPromise>(recursively: true);
ObjectLayoutInspector.TypeLayout.PrintLayout<DelayPromiseWithoutCancellation>(recursively: true);
```
```
Type layout for 'DelayPromise'
Size: 88 bytes. Paddings: 7 bytes (%7 of empty space)
|==============================================================|
| Object Header (8 bytes)                                      |
|--------------------------------------------------------------|
| Method Table Ptr (8 bytes)                                   |
|==============================================================|
|   0-7: Delegate m_action (8 bytes)                           |
|--------------------------------------------------------------|
|  8-15: Object m_stateObject (8 bytes)                        |
|--------------------------------------------------------------|
| 16-23: TaskScheduler m_taskScheduler (8 bytes)               |
|--------------------------------------------------------------|
| 24-31: Object m_continuationObject (8 bytes)                 |
|--------------------------------------------------------------|
| 32-39: ContingentProperties m_contingentProperties (8 bytes) |
|--------------------------------------------------------------|
| 40-43: Int32 m_taskId (4 bytes)                              |
|--------------------------------------------------------------|
| 44-47: Int32 m_stateFlags (4 bytes)                          |
|--------------------------------------------------------------|
|    48: VoidTaskResult m_result (1 byte)                      |
|--------------------------------------------------------------|
| 49-55: padding (7 bytes)                                     |
|--------------------------------------------------------------|
| 56-63: TimerQueueTimer Timer (8 bytes)                       |
|--------------------------------------------------------------|
| 64-71: CancellationToken Token (8 bytes)                     |
|--------------------------------------------------------------|
| 72-87: CancellationTokenRegistration Registration (16 bytes) |
| |=====================================|                      |
| |   0-7: CallbackNode _node (8 bytes) |                      |
| |-------------------------------------|                      |
| |  8-15: Int64 _id (8 bytes)          |                      |
| |=====================================|                      |
|==============================================================|
```
To
```
Type layout for 'DelayPromiseWithoutCancellation'
Size: 64 bytes. Paddings: 7 bytes (%10 of empty space)
|==============================================================|
| Object Header (8 bytes)                                      |
|--------------------------------------------------------------|
| Method Table Ptr (8 bytes)                                   |
|==============================================================|
|   0-7: Delegate m_action (8 bytes)                           |
|--------------------------------------------------------------|
|  8-15: Object m_stateObject (8 bytes)                        |
|--------------------------------------------------------------|
| 16-23: TaskScheduler m_taskScheduler (8 bytes)               |
|--------------------------------------------------------------|
| 24-31: Object m_continuationObject (8 bytes)                 |
|--------------------------------------------------------------|
| 32-39: ContingentProperties m_contingentProperties (8 bytes) |
|--------------------------------------------------------------|
| 40-43: Int32 m_taskId (4 bytes)                              |
|--------------------------------------------------------------|
| 44-47: Int32 m_stateFlags (4 bytes)                          |
|--------------------------------------------------------------|
|    48: VoidTaskResult m_result (1 byte)                      |
|--------------------------------------------------------------|
| 49-55: padding (7 bytes)                                     |
|--------------------------------------------------------------|
| 56-63: TimerQueueTimer Timer (8 bytes)                       |
|==============================================================|
```

/cc @stephentoub

### CODE REVIEW

**Comment by benaadams** [comment]
@dotnet test OSX10.12 x64 Checked Innerloop Build and Test

File: src/System.Private.CoreLib/shared/System/Threading/Tasks/Task.cs [hunk 1]
```diff
@@ -5434,6 +5444,42 @@ internal void Complete()
...
+                setSucceeded = TrySetResult(default);
+
+                // If we set the value, also clean up.
+                if (setSucceeded)
+                {
+                    Timer?.Close();
+                }
+            }
+        }
         #endregion
```
[the remark this answers is not in the document]
Comment (benaadams) [reply src/System.Private.CoreLib/shared/System/Threading/Tasks/Task.cs]: A little, would look like this https://github.com/dotnet/coreclr/pull/22234

File: src/System.Private.CoreLib/shared/System/Threading/Tasks/Task.cs [hunk 2]
```diff
@@ -5369,23 +5369,33 @@ public static Task Delay(int millisecondsDelay, CancellationToken cancellationTo
...
+                if (millisecondsDelay != Timeout.Infinite) // no need to create the timer if it's an infinite timeout
+                {
+                    promise.Timer = new TimerQueueTimer(state => ((DelayPromise)state).Complete(), promise, (uint)millisecondsDelay, Timeout.UnsignedInfinite, flowExecutionContext: false);
+                }
+                // Return the timer proxy task
+                return promise;
             }
+            else
+            {
+                // As above but useing a Delay promise that doesn't support cancellation.
```
Comment (jkotas) [inline src/System.Private.CoreLib/shared/System/Threading/Tasks/Task.cs]:
```suggestion
                // As above but using a Delay promise that doesn't support cancellation.
```

**Comment by Wraith2** [comment]
That 7 bytes of padding bothers me. Is there a way to mask the voidresult into the state flags?

**Comment by stephentoub** [comment]
>  Is there a way to mask the voidresult into the state flags?

No. That's just the generic TResult. The right answer there is to move the promise functionality down from the generic task into the non-generic one. It's on my to-do list to explore that.

**Comment by benaadams** [comment]
> > Is there a way to mask the voidresult into the state flags?

> No. That's just the generic TResult. The right answer there is to move the promise functionality down from the generic task into the non-generic one. It's on my to-do list to explore that.

Could move `m_taskId` into `m_contingentProperties`, then it would drop to 3 bytes ;)

Though it would go back to 4 bytes after making it non-generic (e.g. moving `m_taskId` would be a wash at that point)

**Comment by Wraith2** [comment]
I just hadn't realised it was `T result`, it isn't really wasted padding.

File: src/System.Private.CoreLib/shared/System/Threading/Tasks/Task.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (stephentoub) [reply src/System.Private.CoreLib/shared/System/Threading/Tasks/Task.cs]: Leaving note here for posterity: https://github.com/dotnet/coreclr/pull/22234#issuecomment-457923759

**Comment by stephentoub** [comment]
> The right answer there is to move the promise functionality down from the generic task into the non-generic one. It's on my to-do list to explore that.

https://github.com/dotnet/coreclr/pull/22238

**Comment by benaadams** [comment]
Fixed commits; change should be the same

**Comment by benaadams** [comment]
Though I think I have them reversed O_o

**Comment by benaadams** [comment]
@dotnet-bot test OSX10.12 x64 Checked Innerloop Build and Test

**Comment by benaadams** [comment]
@dotnet-bot test OSX10.12 x64 Checked Innerloop Build and Test

**Comment by jkotas** [comment]
@benaadams Could you please resolve the conflict?

**Comment by stephentoub** [comment]
Resolved.

**Pull request merged**